feat(event-gateway): Virtual clusters (#3095)
Kong/developer.konghq.com: app/event-gateway/entities/virtual-cluster.md

  - text: "Listeners"
    url: /event-gateway/entities/listeners/

tools:
    - konnect-api
    - terraform
tags:
  - policy

works_on:
  - konnect

schema:
    api: event-gateway/knep
    path: /schemas/VirtualCluster

api_specs:
    - event-gateway/knep

products:
    - event-gateway
---

## What is a virtual cluster?

Virtual clusters are the primary way clients interact with the {{site.event_gateway_short}} proxy. 
They allow you to isolate clients from each other when connecting to the same [backend cluster](/event-gateway/entities/backend-clusters/), 
and provide each client with modified view while still appearing as a standard Kafka cluster.

Here's how it works:
1. The Kafka client produces an event.
1. A listener forwards it to the correct virtual cluster.
1. The virtual cluster applies policies and proxies the modified event data to the backend cluster.
1. The backend cluster, representing a Kafka cluster, receives data and sends a response.

{% mermaid %}
flowchart LR
    A[Kafka client] --> B[Listener
    + listener policies]
    B --> C[Virtual cluster
    + consume, produce, and cluster policies]
    C --> D[Backend 
    cluster]

    style C stroke:#86e2cc

> {% endmermaid %}

Reviewer: Can we do another arrow from the backend cluster so that it shows it sending a response? For now it looks like it just sends the event data to the backend cluster and it stops.

{:.info}
> **Note**: Each virtual cluster can only expose one backend cluster, but you can have multiple virtual clusters connected to one backend.

## Why use a virtual cluster?

Virtual clusters let you apply governance and security features to event streams.
This way, a single Kafka cluster can to be sliced into multiple endpoints, each with its own security policy.

<!--vale off-->
{% table %}
columns:
  - title: Use case
    key: use_case
  - title: Description
    key: description
rows:
  - use_case: "Policy enforcement"
    description: |
      Define policies on virtual clusters to govern client behavior. Policies include transformations, filtering, enforcing encryption and decryption standards, access control, and more.

>   - use_case: "Authentication and mediation"
>     description: |
>       Manage client authentication to the proxy with authentication mediation. 
>       {{site.event_gateway_short}} can validate client credentials (like an OAuth token) before using separate credentials to connect to the upstream backend cluster.

Reviewer: ```suggestion
  - use_case: "Authentication and mediation"
    description: |
    Use authentication mediation to control access between clients and backend clusters. 
    {{site.event_gateway_short}} authenticates clients (for example, with OAuth tokens) 
    and re-authenticates separately when forwarding requests to the backend.
```

Is this right?

  - use_case: "Topic and cluster virtualization"
    description: |
      Use topic and cluster virtualization to simplify change management and security. Virtual clusters can expose only a subset of topics on the backend cluster.
  
  - use_case: "Namespacing and topic rewriting"
    description: |
      Virtual clusters support Namespaces, which rewrite and enforce consistent prefixes for topic and consumer group names, exposing specific topics and consumer groups. 
      For example, a virtual cluster might expose a topic named `orders`, which internally maps to a physical topic like `dev-orders` or `prod-orders` on the backend cluster.
  

>   - use_case: Cost optimization

Reviewer: Just calling out that this use case doesn't feel like the other ones.

Here are some examples:

* **Environment isolation**: You can create isolated `dev`, `test`, and `prod` namespaces on top of the same physical Kafka cluster.
If you have a topic named `orders` in each virtual cluster, it can map to different backend topics: `dev-orders`, `test-orders`, and `prod-orders`. 
This provides isolation and automatic name resolution per environment.

* **External partner isolation**: You can expose the same backend topic to different external partners with data filtering. 
For instance, a single `orders` topic can be exposed through separate virtual clusters (`customer-a`, `customer-b`, `customer-c`), with each customer seeing only their own orders.

* **Reverse mapping**: One backend topic (`orders`) can appear as multiple separate topics (`dev-orders`, `test-orders`, `prod-orders`) across different virtual clusters, each pre-filtered for specific users.

## Authentication

Authentication on the virtual cluster is used to authenticate clients to the proxy. 
The virtual cluster supports multiple authentication methods and can mediate authentication between clients and backend clusters.

> The following auth methods are supported:

Reviewer: ```suggestion
Supported methods:
```

---
title: "Virtual clusters"
content_type: reference
layout: gateway_entity

description: |
    Virtual clusters are {{site.event_gateway_short}} entities that expose a modified view of the backend cluster to clients.
related_resources:
  - text: "{{site.event_gateway}} Policy Hub"
    url: /event-gateway/policies/
  - text: "Policies"
    url: /event-gateway/entities/policies/
  - text: "Backend Clusters"
    url: /event-gateway/entities/backend-clusters/
  - text: "Listeners"
    url: /event-gateway/entities/listeners/

tools:
    - konnect-api
    - terraform
tags:
  - policy

works_on:
  - konnect

schema:
    api: event-gateway/knep
    path: /schemas/VirtualCluster

api_specs:
    - event-gateway/knep

products:
    - event-gateway
---

## What is a virtual cluster?

Virtual clusters are the primary way clients interact with the {{site.event_gateway_short}} proxy. 
They allow you to isolate clients from each other when connecting to the same [backend cluster](/event-gateway/entities/backend-clusters/), 
and provide each client with modified view while still appearing as a standard Kafka cluster.

> Here's how it works:

Reviewer: ```suggestion
The virtual cluster workflow operates as follows:
```
